### PR TITLE
Add DEN algorithm

### PR DESCRIPTION
**Pull request by collinleiber**
- Add the Deep Embedded Network for Clustering algorithm
- Change default implementation of MSE
- Change implementations of Deep Clustering algorithms to comply with sklearn check_estimator

### CODE REVIEW

File: clustpy/deep/_train_utils.py [hunk 1]
```diff
@@ -45,13 +46,17 @@ def _get_neural_network(input_dim: int, embedding_size: int = 10, neural_network
...
         Can also be None. In this case a new neural network will be created using neural_network_class and neural_network_params (default: None)
     neural_network_class : torch.nn.Module
         The neural network class that should be used (default: FeedforwardAutoencoder)
     neural_network_params : dict
         Parameters to be used when creating a new neural network using the neural_network_class (default: None)
     neural_network_weights : str
         Path to a file containing the state_dict of the neural_network (default: None)
+    device : torch.device
+        The device on which to perform the computations.
+        If device is None then it will be automatically chosen: if a gpu is available the gpu with the highest amount of free memory will be chosen (default: None)
```
Comment (Lumik7) [inline clustpy/deep/_train_utils.py:59]: that's a nice addition

File: clustpy/deep/den.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+from sklearn.cluster import KMeans
+from collections.abc import Callable
+from clustpy.utils.checks import check_parameters
+
+
+class DEN(_AbstractDeepClusteringAlgo):
+    """
+    The Deep Embedding Network (DEN) algorithm.
+    It trains a neural network by optimizing a loss functions consisting of three components.
+    These are (1) the standrad loss function of the neural netork (e.g. construction loss for autoencoders), (2) the locality-preserving constraint and (3) the group sparsity constraint.
```
Comment (Lumik7) [inline clustpy/deep/den.py:23]: construction loss --> reconstruction loss

**Review by Lumik7 — APPROVED** [review]
Some minor changes are needed, otherwise it can be merged
